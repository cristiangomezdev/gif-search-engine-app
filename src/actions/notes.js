import { db, collection } from "../firebase/firebaseConfig";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { doc, updateDoc, addDoc, deleteDoc } from "@firebase/firestore"
import Swal from 'sweetalert2';
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await addDoc(collection(db, `${uid}/journal/notes`), newNote)
    dispatch(activeNote(doc.id, newNote))
    dispatch(addNewNote(doc.id,newNote))

  }
}

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note
  }
})
export const addNewNote = (id,note) => ({
  type:types.notesAddNew,
  payload:{
    id, ...note 
  }
})
export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes
})

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  }
}
export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth

    if (!note.url) {
      delete note.url
    }

    const noteToFirestore = { ...note }
    delete noteToFirestore.id
    const noteRef = doc(db, `${uid}/journal/notes/${note.id}`)
    await updateDoc(noteRef, noteToFirestore);
    Swal.fire('Saved', note.title, 'success');

    dispatch(refreshNote(note.id, note))
  }
}

export const refreshNote = (id, note) => ({
  type: types.notesUpdated,
  payload: { id, note: { id, ...note } }
})

export const StartUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    const fileUrl = await fileUpload(file);
    Swal.fire({ title: 'Uploading', text: 'Please wait...', showCancelButton: false, showConfirmButton: false, willOpen: () => { Swal.showLoading() } });
    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote))
    Swal.close();
  }
}

export const startDeleting = (id) =>{
  return async(dispatch,getState)=>{

    const uid = getState().auth.uid;
    const noteRef = doc(db, `${uid}/journal/notes/${id}`)
    await deleteDoc(noteRef);

    dispatch(deleteNote(id));
  }
}
export const deleteNote = (id) =>({
  type:types.notesDelete,
  payload:id
})

export const noteLogout = () =>({
  type:types.notesLogoutCleaning
})