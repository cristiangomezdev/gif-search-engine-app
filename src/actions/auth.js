import { googleAuthProvider } from '../firebase/firebaseConfig';
import { createUserWithEmailAndPassword,signOut, updateProfile,  getAuth ,signInWithEmailAndPassword , signInWithPopup} from 'firebase/auth';
import { startLoading,finishLoading } from './ui';
import {types} from '../types/types';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading())
        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
            .then(({user}) => {
                dispatch(
                    login( user.uid, user.displayName )
                );
                dispatch(finishLoading())

            })
            .catch((error) => {

                Swal.fire('Error',error.message,'error')
            });
    };
};

export const startRegisterWithEmailPasswordName = (email,password,name) => {
    return (dispatch) => {
 
        const auth = getAuth();
            createUserWithEmailAndPassword(auth, email, password)
                .then( async({ user }) => {
 
                await updateProfile( user, { displayName: name });
 

        })
        .catch(error =>{Swal.fire('Error',error.message,'error')})
    }

}

export const startGoogleLogin = () =>{
    return(dispatch) =>{
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then(({user}) =>{
                dispatch(login(user.uid, user.displayName))
            });
    }
}
export const login = (uid,displayName) =>{
    return{
        type:types.login,
        payload:{
            uid,
            displayName
        }
    }
}
export const logoutAction = () => {
    return async(dispatch) => {
        const auth = getAuth();
        await signOut(auth);
        dispatch(logout());
        dispatch(noteLogout());
    }
}

export const logout = ()=>({
    type:types.logout
})