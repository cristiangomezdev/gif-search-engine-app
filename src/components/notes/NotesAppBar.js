import React from 'react'
import { startSaveNote, StartUploading } from '../../actions/notes'
import {useDispatch,useSelector} from 'react-redux';
import moment from 'moment';

export const NotesAppBar = () => {

        
        const dispatch = useDispatch();
        const {active} = useSelector(state =>state.notes)
        const noteDate = moment(active.note)
        const handleSave= () =>{
            dispatch(startSaveNote(active))
        }
        const handlePictureClick = () =>{
            document.querySelector('#fileSelector').click();
        }
        const handleFileChange = (e)=>{
            const file = e.target.files[0]
            if(file){
                dispatch(StartUploading(file));
            }
        }
    return (
        <div className="notes__appbar">
            <span>Date: {noteDate.format('MMMM Do YYYY')}</span>
            <input id="fileSelector" name="file" type="file" style={{display:'none'}} onChange={handleFileChange} />
            <div>
                <button className="btn" onClick={handlePictureClick}> photo</button>
                <button className="btn" onClick={handleSave}> save</button>
            </div>
        </div>
    )
}
