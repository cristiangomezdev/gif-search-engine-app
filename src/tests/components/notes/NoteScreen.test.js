import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';


import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { activeNote } from '../../../actions/notes';
import { NoteScreen } from '../../../components/notes/NoteScreen';



jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}));


const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        uid: '1',
        name: 'cris'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {       
                id: 1234,
                title: 'hola',
                body: 'mundo',
                date: 0
        },
        notes: []
    }
};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <NoteScreen />
    </Provider>
)
describe('TEST notescreen', () => {
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('debe de disparar el active note', () => {
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'hola de nuevo'
            }
        })
        expect(activeNote).toHaveBeenLastCalledWith(
            1234,
            {
                body: 'mundo',
                title: 'hola de nuevo',
                id:1234,
                date: 0
            }
        )
});
});