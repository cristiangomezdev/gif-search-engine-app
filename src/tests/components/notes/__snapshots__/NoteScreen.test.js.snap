// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST notescreen Debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 of august 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
         picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
         save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="some awesome title"
      type="text"
      value="hola"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="what happened today"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
